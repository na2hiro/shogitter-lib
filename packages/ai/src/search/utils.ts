import { ShogitterAiException } from "../ShogitterAiException.js";

export function randomSelect<T>(arr: T[]) {
  if (arr.length === 0)
    throw new ShogitterAiException("randomSelect: empty array");
  return arr[Math.floor(Math.random() * arr.length)];
}
