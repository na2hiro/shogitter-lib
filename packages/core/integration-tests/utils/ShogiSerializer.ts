import Shogi from "../../src/Shogi";

export const serialize = (
  val: Shogi,
  config: any,
  indentation: any,
  depth: any,
  refs: any,
  printer: any
) => {
  const { ban, kifu, players, ...obj } = val.getObject();

  return `ban:
(1) ${prettifyMochigoma(players[1].mochigoma)}${
    players[3] ? `\n(3) ${prettifyMochigoma(players[3].mochigoma)}` : ""
  }
${val.ban.__toString()}(0) ${prettifyMochigoma(players[0].mochigoma)}${
    players[2] ? `\n(2) ${prettifyMochigoma(players[2].mochigoma)}` : ""
  }
kifu:
${val.kifu.arrayKifu
  .map((line) => `${line.disp} ${JSON.stringify(line.move)}`)
  .join("\n")}
${JSON.stringify(obj, null, 2)}`;
};

export const test = (val: any) => {
  return val instanceof Shogi;
};

const prettifyMochigoma = (map: { [species: string]: number }) => {
  let ret = "";
  for (let species in map) {
    ret += species + map[species] + " ";
  }
  return ret;
};
