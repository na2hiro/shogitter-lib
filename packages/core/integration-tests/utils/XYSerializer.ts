import XY from "../../src/XY";

export const serialize = (
  val: XY,
  config: any,
  indentation: any,
  depth: any,
  refs: any,
  printer: any
) => {
  return val.toString();
};

export const test = (val: any) => {
  return val instanceof XY;
};
