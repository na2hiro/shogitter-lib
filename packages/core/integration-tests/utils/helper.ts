import { Shogi, XY } from "../../src";

export function getShogiHelper(ruleId: number) {
  const shogi = Shogi.ofRuleId(ruleId);
  shogi.start();
  return {
    shogi,
    move: (
      fromX: number,
      fromY: number,
      toX: number,
      toY: number,
      nari = false
    ) => {
      shogi.move(new XY(fromX, fromY), new XY(toX, toY), nari);
    },
    put: (toX: number, toY: number, species: string) => {
      shogi.put(new XY(toX, toY), species, shogi.teban.get());
    },
  };
}
