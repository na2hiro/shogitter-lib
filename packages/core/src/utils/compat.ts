export function deepCopy<T>(obj: T): T {
  return typeof structuredClone !== "undefined"
    ? structuredClone(obj)
    : JSON.parse(JSON.stringify(obj));
}
