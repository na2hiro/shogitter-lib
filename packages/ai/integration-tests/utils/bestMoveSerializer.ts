import { SnapshotSerializer } from "vitest";
import { BestMove } from "../../src/search/minimax";
import { Move } from "@shogitter/core/src";

export default {
  serialize(val: BestMove<Move>, config, indentation, depth, refs, printer) {
    // `printer` is a function that serializes a value using existing plugins.
    return `${printer(printBestMove(val), config, indentation, depth, refs)}`;
  },
  test(val) {
    return (
      val &&
      Object.prototype.hasOwnProperty.call(val, "moves") &&
      Object.prototype.hasOwnProperty.call(val, "score")
    );
  },
} satisfies SnapshotSerializer;

function printMove(move: Move) {
  if (move.type === "move") {
    return `${move.from[0]}${move.from[1]}${move.to[0]}${move.to[1]}${
      move.nari ? "+" : ""
    }`;
  } else if (move.type == "put") {
    return `${move.to[0]}${move.to[1]}${move.put}`;
  } else {
    return move.type;
  }
}

function printBestMove(val: BestMove<Move>) {
  return (
    val.moves.map(printMove).join(" ") +
    ` ${val.score}` +
    (val.trace?.length > 0
      ? "\n" + val.trace?.map(printBestMove).join("\n")
      : "")
  );
}
