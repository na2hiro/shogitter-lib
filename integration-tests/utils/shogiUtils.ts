import Shogi from "../../src/Shogi";
import XY from "../../src/XY";
import {Species} from "../../src/Ban";
import {serialize} from "./ShogiSerializer";
import {Direction} from "../../src/Direction";

export function move(shogi: Shogi, fromX: number, fromY: number, toX: number, toY: number, nari = false) {
    try {
        rawMove(shogi, fromX, fromY, toX, toY, nari);
    } catch (e) {
        console.error(serialize(shogi, null, null, null, null, null))
        throw e;
    }
}

export function rawMove(shogi: Shogi, fromX: number, fromY: number, toX: number, toY: number, nari = false) {
    shogi.runCommand({
        type: "move",
        from: [fromX, fromY],
        to: [toX, toY],
        nari
    })
}

export function put(shogi: Shogi, toX: number, toY: number, species: Species, id?: number) {
    try {
        rawPut(shogi, toX, toY, species, id);
    } catch (e) {
        console.error(serialize(shogi, null, null, null, null, null))
        throw e;
    }
}

export function rawPut(shogi: Shogi, toX: number, toY: number, species: Species, id?: number) {
    shogi.runCommand({
        type: "put",
        to: [toX, toY],
        put: species,
        direction: shogi.teban.getNowDirection(),
        id
    })
}

export function assertCellIsNull(shogi: Shogi, x: number, y: number) {
    expect(shogi.ban.get(new XY(x, y)).isNull()).toBeTruthy;
}

export function assertCell(shogi: Shogi, x: number, y: number, direction: Direction, species: Species) {
    const koma = shogi.ban.get(new XY(x, y));
    expect(koma.isNull()).toBeFalsy();
    expect(koma.species).toBe(species);
    expect(koma.direction).toBe(direction);
}
