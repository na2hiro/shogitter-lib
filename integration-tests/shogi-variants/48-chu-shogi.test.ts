import Shogi from "../../src/Shogi";
import { move, rawMove } from "../utils/shogiUtils";

describe("Chu shogi", () => {
  let shogi: Shogi;
  beforeEach(() => {
    shogi = Shogi.ofRuleId(48);
    shogi.start();
  });
  it("shishi can move", () => {
    move(shogi, 7, 10, 7, 8);
    move(shogi, 6, 3, 6, 5);
    expect(shogi).toMatchSnapshot();
  });
  it("cannot take shishi right after capturing shishi (Senjishi)", () => {
    move(shogi, 8, 9, 8, 8);
    move(shogi, 5, 4, 5, 5);
    move(shogi, 9, 8, 9, 7);
    move(shogi, 4, 5, 4, 6);
    move(shogi, 7, 10, 9, 8);
    move(shogi, 6, 3, 4, 5);
    move(shogi, 9, 10, 4, 5);
    expect(shogi).toMatchSnapshot();
    expect(() => rawMove(shogi, 4, 3, 9, 8)).toThrow();
  });
  it("can take shishi next to shishi", () => {
    move(shogi, 7, 10, 7, 8);
    move(shogi, 6, 3, 8, 5);
    move(shogi, 7, 8, 9, 7);
    move(shogi, 8, 5, 8, 6);
    move(shogi, 8, 6, 9, 6);
    move(shogi, 9, 7, 9, 6);
    expect(shogi).toMatchSnapshot();
  });
  it("can take shishi 2 pieces away which doesn't have kiki", () => {
    move(shogi, 7, 10, 7, 8);
    move(shogi, 6, 3, 7, 5);
    move(shogi, 7, 8, 5, 7);
    move(shogi, 7, 5, 5, 7);
    expect(shogi).toMatchSnapshot();
  });
  it("cannot take shishi 2 pieces away which has kiki", () => {
    move(shogi, 7, 10, 8, 8);
    move(shogi, 6, 3, 8, 5);
    move(shogi, 8, 8, 8, 7);
    move(shogi, 8, 7, 9, 7);
    expect(shogi).toMatchSnapshot();
    expect(() => rawMove(shogi, 8, 5, 9, 7)).toThrow();
  });
  it("cannot take shishi 2 pieces away which has kiki through the shishi to move", () => {
    move(shogi, 7, 10, 8, 8);
    move(shogi, 9, 5, 9, 6);
    move(shogi, 8, 8, 9, 6);
    move(shogi, 9, 4, 9, 5);
    move(shogi, 9, 6, 10, 6);
    move(shogi, 10, 6, 9, 5);
    move(shogi, 6, 3, 6, 5);
    move(shogi, 2, 9, 2, 8);
    move(shogi, 6, 5, 7, 7);
    move(shogi, 5, 9, 5, 8);
    expect(shogi).toMatchSnapshot();
    expect(() => rawMove(shogi, 7, 7, 9, 5)).toThrow();
  });
  it("cannot tsukegui with chunin", () => {
    move(shogi, 7, 10, 8, 8);
    move(shogi, 6, 3, 8, 5);
    move(shogi, 8, 8, 8, 7);
    move(shogi, 8, 7, 9, 7);
    move(shogi, 9, 5, 9, 6);
    move(shogi, 9, 7, 9, 6);
    expect(shogi).toMatchSnapshot();
    expect(() => rawMove(shogi, 9, 6, 8, 5)).toThrow();
  });
  it("can tsukegui", () => {
    move(shogi, 7, 10, 8, 8);
    move(shogi, 6, 3, 8, 5);
    move(shogi, 5, 9, 5, 8);
    move(shogi, 8, 5, 7, 5);
    move(shogi, 7, 5, 7, 6);
    move(shogi, 4, 10, 7, 7);
    move(shogi, 7, 6, 7, 7);
    move(shogi, 7, 7, 8, 8);
    move(shogi, 8, 9, 8, 8);
    expect(shogi).toMatchSnapshot();
  });
});
