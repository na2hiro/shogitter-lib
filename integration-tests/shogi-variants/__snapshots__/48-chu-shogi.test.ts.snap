// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chu shogi can take shishi 2 pieces away which doesn't have kiki 1`] = `
ban:
(1) 
1ab1cl1ci1ad1ae1cf1ah1ae1ad1ci1cl1ab
1cm___1af___1cn1cp1co1cn___1af___1cm
1cr1cs1ag1am1an1ct___1an1am1ag1cs1cr
1aa1aa1aa1aa1aa1aa1aa1aa1aa1aa1aa1aa
_________1cq____________1cq_________
____________________________________
_____________________1bl____________
_________0cq____________0cq_________
0aa0aa0aa0aa0aa0aa0aa0aa0aa0aa0aa0aa
0cr0cs0ag0am0an___0ct0an0am0ag0cs0cr
0cm___0af___0cn0co0cp0cn___0af___0cm
0ab0cl0ci0ad0ae0ah0cf0ae0ad0ci0cl0ab
(0) 
kifu:
☗7八獅子 [0,[7,8,[],[0,"bl"]],[7,10,[0,"bl"],[]]]
☖7五獅子 [1,[7,5,[],[1,"bl"]],[6,3,[1,"bl"],[]]]
☗5七獅子 [0,[5,7,[],[0,"bl"]],[7,8,[0,"bl"],[]]]
☖同獅子 [1,[5,7,[0,"bl"],[1,"bl"]],[7,5,[1,"bl"],[]]]
{
  "version": "0.0",
  "status": {
    "num": 1
  },
  "ruleid": 48,
  "teban": 0,
  "turn": 0,
  "date": {},
  "moving": null
}
`;

exports[`Chu shogi can take shishi next to shishi 1`] = `
ban:
(1) 
1ab1cl1ci1ad1ae1cf1ah1ae1ad1ci1cl1ab
1cm___1af___1cn1cp1co1cn___1af___1cm
1cr1cs1ag1am1an1ct___1an1am1ag1cs1cr
1aa1aa1aa1aa1aa1aa1aa1aa1aa1aa1aa1aa
_________1cq____________1cq_________
_________0bl________________________
____________________________________
_________0cq____________0cq_________
0aa0aa0aa0aa0aa0aa0aa0aa0aa0aa0aa0aa
0cr0cs0ag0am0an___0ct0an0am0ag0cs0cr
0cm___0af___0cn0co0cp0cn___0af___0cm
0ab0cl0ci0ad0ae0ah0cf0ae0ad0ci0cl0ab
(0) 
kifu:
☗7八獅子 [0,[7,8,[],[0,"bl"]],[7,10,[0,"bl"],[]]]
☖8五獅子 [1,[8,5,[],[1,"bl"]],[6,3,[1,"bl"],[]]]
☗9七獅子 [0,[9,7,[],[0,"bl"]],[7,8,[0,"bl"],[]]]
☖8六獅子 [1,[8,6,[],[1,"bl"]],[8,5,[1,"bl"],[]]]
☖9六獅子 [1,[9,6,[],[1,"bl"]],[8,6,[1,"bl"],[]]]
☗同獅子 [0,[9,6,[1,"bl"],[0,"bl"]],[9,7,[0,"bl"],[]]]
{
  "version": "0.0",
  "status": {
    "num": 1
  },
  "ruleid": 48,
  "teban": 0,
  "turn": 0,
  "date": {},
  "moving": {
    "xy": [
      9,
      6
    ],
    "status": 0
  }
}
`;

exports[`Chu shogi can tsukegui 1`] = `
ban:
(1) 
1ab1cl1ci1ad1ae1cf1ah1ae1ad1ci1cl1ab
1cm___1af___1cn1cp1co1cn___1af___1cm
1cr1cs1ag1am1an1ct___1an1am1ag1cs1cr
1aa1aa1aa1aa1aa1aa1aa1aa1aa1aa1aa1aa
_________1cq____________1cq_________
____________________________________
____________________________________
_________0cq0aa______0aa0cq_________
0aa0aa0aa0aa___0aa0aa___0aa0aa0aa0aa
0cr0cs0ag0am0an___0ct0an___0ag0cs0cr
0cm___0af___0cn0co0cp0cn___0af___0cm
0ab0cl0ci0ad0ae0ah0cf0ae0ad0ci0cl0ab
(0) 
kifu:
☗8八獅子 [0,[8,8,[],[0,"bl"]],[7,10,[0,"bl"],[]]]
☖8五獅子 [1,[8,5,[],[1,"bl"]],[6,3,[1,"bl"],[]]]
☗5八歩 [0,[5,8,[],[0,"aa"]],[5,9,[0,"aa"],[]]]
☖7五獅子 [1,[7,5,[],[1,"bl"]],[8,5,[1,"bl"],[]]]
☖7六獅子 [1,[7,6,[],[1,"bl"]],[7,5,[1,"bl"],[]]]
☗7七馬 [0,[7,7,[],[0,"am"]],[4,10,[0,"am"],[]]]
☖同獅子 [1,[7,7,[0,"am"],[1,"bl"]],[7,6,[1,"bl"],[]]]
☖8八獅子 [1,[8,8,[0,"bl"],[1,"bl"]],[7,7,[1,"bl"],[]]]
☗同歩 [0,[8,8,[1,"bl"],[0,"aa"]],[8,9,[0,"aa"],[]]]
{
  "version": "0.0",
  "status": {
    "num": 1
  },
  "ruleid": 48,
  "teban": 1,
  "turn": 0,
  "date": {},
  "moving": null
}
`;

exports[`Chu shogi cannot take shishi 2 pieces away which has kiki 1`] = `
ban:
(1) 
1ab1cl1ci1ad1ae1cf1ah1ae1ad1ci1cl1ab
1cm___1af___1cn1cp1co1cn___1af___1cm
1cr1cs1ag1am1an1ct___1an1am1ag1cs1cr
1aa1aa1aa1aa1aa1aa1aa1aa1aa1aa1aa1aa
_________1cq1bl_________1cq_________
____________________________________
_________0bl________________________
_________0cq____________0cq_________
0aa0aa0aa0aa0aa0aa0aa0aa0aa0aa0aa0aa
0cr0cs0ag0am0an___0ct0an0am0ag0cs0cr
0cm___0af___0cn0co0cp0cn___0af___0cm
0ab0cl0ci0ad0ae0ah0cf0ae0ad0ci0cl0ab
(0) 
kifu:
☗8八獅子 [0,[8,8,[],[0,"bl"]],[7,10,[0,"bl"],[]]]
☖8五獅子 [1,[8,5,[],[1,"bl"]],[6,3,[1,"bl"],[]]]
☗8七獅子 [0,[8,7,[],[0,"bl"]],[8,8,[0,"bl"],[]]]
☗9七獅子 [0,[9,7,[],[0,"bl"]],[8,7,[0,"bl"],[]]]
{
  "version": "0.0",
  "status": {
    "num": 1
  },
  "ruleid": 48,
  "teban": 1,
  "turn": 0,
  "date": {},
  "moving": null
}
`;

exports[`Chu shogi cannot take shishi 2 pieces away which has kiki through the shishi to move 1`] = `
ban:
(1) 
1ab1cl1ci1ad1ae1cf1ah1ae1ad1ci1cl1ab
1cm___1af___1cn1cp1co1cn___1af___1cm
1cr1cs1ag1am1an1ct___1an1am1ag1cs1cr
1aa1aa1aa___1aa1aa1aa1aa1aa1aa1aa1aa
_________0bl____________1cq_________
____________________________________
_______________1bl__________________
_________0cq_________0aa0cq___0aa___
0aa0aa0aa0aa0aa0aa0aa___0aa0aa___0aa
0cr0cs0ag0am0an___0ct0an0am0ag0cs0cr
0cm___0af___0cn0co0cp0cn___0af___0cm
0ab0cl0ci0ad0ae0ah0cf0ae0ad0ci0cl0ab
(0) 
kifu:
☗8八獅子 [0,[8,8,[],[0,"bl"]],[7,10,[0,"bl"],[]]]
☖9六仲人 [1,[9,6,[],[1,"cq"]],[9,5,[1,"cq"],[]]]
☗同獅子 [0,[9,6,[1,"cq"],[0,"bl"]],[8,8,[0,"bl"],[]]]
☖9五歩 [1,[9,5,[],[1,"aa"]],[9,4,[1,"aa"],[]]]
☗10六獅子 [0,[10,6,[],[0,"bl"]],[9,6,[0,"bl"],[]]]
☗9五獅子 [0,[9,5,[1,"aa"],[0,"bl"]],[10,6,[0,"bl"],[]]]
☖6五獅子 [1,[6,5,[],[1,"bl"]],[6,3,[1,"bl"],[]]]
☗2八歩 [0,[2,8,[],[0,"aa"]],[2,9,[0,"aa"],[]]]
☖7七獅子 [1,[7,7,[],[1,"bl"]],[6,5,[1,"bl"],[]]]
☗5八歩 [0,[5,8,[],[0,"aa"]],[5,9,[0,"aa"],[]]]
{
  "version": "0.0",
  "status": {
    "num": 1
  },
  "ruleid": 48,
  "teban": 1,
  "turn": 0,
  "date": {},
  "moving": null
}
`;

exports[`Chu shogi cannot take shishi right after capturing shishi (Senjishi) 1`] = `
ban:
(1) 
1ab1cl1ci1ad1ae1cf1ah1ae1ad1ci1cl1ab
1cm___1af___1cn1cp1co1cn___1af___1cm
1cr1cs1ag1am1an1ct___1an1am1ag1cs1cr
1aa1aa1aa1aa1aa1aa1aa___1aa1aa1aa1aa
_________1cq_________1aa0am_________
________________________1cq_________
_________0cq________________________
_________0bl0aa_________0cq_________
0aa0aa0aa0aa___0aa0aa0aa0aa0aa0aa0aa
0cr0cs0ag___0an___0ct0an0am0ag0cs0cr
0cm___0af___0cn0co0cp0cn___0af___0cm
0ab0cl0ci0ad0ae0ah0cf0ae0ad0ci0cl0ab
(0) 
kifu:
☗8八歩 [0,[8,8,[],[0,"aa"]],[8,9,[0,"aa"],[]]]
☖5五歩 [1,[5,5,[],[1,"aa"]],[5,4,[1,"aa"],[]]]
☗9七仲人 [0,[9,7,[],[0,"cq"]],[9,8,[0,"cq"],[]]]
☖4六仲人 [1,[4,6,[],[1,"cq"]],[4,5,[1,"cq"],[]]]
☗9八獅子 [0,[9,8,[],[0,"bl"]],[7,10,[0,"bl"],[]]]
☖4五獅子 [1,[4,5,[],[1,"bl"]],[6,3,[1,"bl"],[]]]
☗同馬 [0,[4,5,[1,"bl"],[0,"am"]],[9,10,[0,"am"],[]]]
{
  "version": "0.0",
  "status": {
    "num": 1
  },
  "ruleid": 48,
  "teban": 1,
  "turn": 0,
  "date": {},
  "moving": null
}
`;

exports[`Chu shogi cannot tsukegui with chunin 1`] = `
ban:
(1) 
1ab1cl1ci1ad1ae1cf1ah1ae1ad1ci1cl1ab
1cm___1af___1cn1cp1co1cn___1af___1cm
1cr1cs1ag1am1an1ct___1an1am1ag1cs1cr
1aa1aa1aa1aa1aa1aa1aa1aa1aa1aa1aa1aa
____________1bl_________1cq_________
_________0bl________________________
____________________________________
_________0cq____________0cq_________
0aa0aa0aa0aa0aa0aa0aa0aa0aa0aa0aa0aa
0cr0cs0ag0am0an___0ct0an0am0ag0cs0cr
0cm___0af___0cn0co0cp0cn___0af___0cm
0ab0cl0ci0ad0ae0ah0cf0ae0ad0ci0cl0ab
(0) 
kifu:
☗8八獅子 [0,[8,8,[],[0,"bl"]],[7,10,[0,"bl"],[]]]
☖8五獅子 [1,[8,5,[],[1,"bl"]],[6,3,[1,"bl"],[]]]
☗8七獅子 [0,[8,7,[],[0,"bl"]],[8,8,[0,"bl"],[]]]
☗9七獅子 [0,[9,7,[],[0,"bl"]],[8,7,[0,"bl"],[]]]
☖9六仲人 [1,[9,6,[],[1,"cq"]],[9,5,[1,"cq"],[]]]
☗同獅子 [0,[9,6,[1,"cq"],[0,"bl"]],[9,7,[0,"bl"],[]]]
{
  "version": "0.0",
  "status": {
    "num": 1
  },
  "ruleid": 48,
  "teban": 0,
  "turn": 0,
  "date": {},
  "moving": {
    "xy": [
      9,
      6
    ],
    "status": 0
  }
}
`;

exports[`Chu shogi shishi can move 1`] = `
ban:
(1) 
1ab1cl1ci1ad1ae1cf1ah1ae1ad1ci1cl1ab
1cm___1af___1cn1cp1co1cn___1af___1cm
1cr1cs1ag1am1an1ct___1an1am1ag1cs1cr
1aa1aa1aa1aa1aa1aa1aa1aa1aa1aa1aa1aa
_________1cq______1bl___1cq_________
____________________________________
____________________________________
_________0cq___0bl______0cq_________
0aa0aa0aa0aa0aa0aa0aa0aa0aa0aa0aa0aa
0cr0cs0ag0am0an___0ct0an0am0ag0cs0cr
0cm___0af___0cn0co0cp0cn___0af___0cm
0ab0cl0ci0ad0ae0ah0cf0ae0ad0ci0cl0ab
(0) 
kifu:
☗7八獅子 [0,[7,8,[],[0,"bl"]],[7,10,[0,"bl"],[]]]
☖6五獅子 [1,[6,5,[],[1,"bl"]],[6,3,[1,"bl"],[]]]
{
  "version": "0.0",
  "status": {
    "num": 1
  },
  "ruleid": 48,
  "teban": 0,
  "turn": 0,
  "date": {},
  "moving": null
}
`;
